app/src/main/java/one/mixin/android/job/BlazeMessageService.kt

```diff
@@ -301,21 +313,79 @@ class BlazeMessageService : LifecycleService(), NetworkEventProvider.Listener, C
         }
     }
 
+    @ExperimentalTime
     private tailrec suspend fun processFloodMessage(): Boolean {
         val messages = floodMessageDao.findFloodMessages()
         return if (!messages.isNullOrEmpty()) {
-            messages.forEach { message ->
-                val data = gson.fromJson(message.data, BlazeMessageData::class.java)
-                if (data.category.startsWith("WEBRTC_") || data.category.startsWith("KRAKEN_")) {
-                    callMessageDecrypt.onRun(data)
-                } else {
-                    messageDecrypt.onRun(data)
-                }
-                floodMessageDao.delete(message)
-            }
+            Timber.d(
+                "process flood messages size: ${messages.size}, cost: ${measureTime {
+                    val ids = messages.map { it.messageId }
+                    val existsMessageIds = database.messageDao().findMessageIdsByIds(ids).toSet()
+                    val existsHistories = database.messageHistoryDao().findMessageHistoryByIds(ids).toSet()
+                    val existsUnion = existsMessageIds.union(existsHistories)
+                    val candidateMessages = messages.filter { existsUnion.contains(it.messageId).not() }
+
+                    processCandidateMessages(candidateMessages, messages, existsHistories)
+                }}"
+            )
             processFloodMessage()
         } else {
             false
         }
     }
+
+    private fun processCandidateMessages(
+        candidateMessages: List<FloodMessage>,
+        messages: List<FloodMessage>,
+        existsHistories: Set<String>,
+    ) {
+        candidateMessages.forEach { message ->
+            val data = gson.fromJson(message.data, BlazeMessageData::class.java)
+            if (data.category.startsWith("WEBRTC_") || data.category.startsWith("KRAKEN_")) {
+                callMessageDecrypt.onRun(data)
+            } else {
+                messageDecrypt.onRun(data)
+            }
+        }
+
+        runInTransaction {
+            val pendingMessages = messageDecrypt.pendingMessages + callMessageDecrypt.pendingMessages
+            val updateUnseenConversationIds = mutableSetOf<String>()
+            pendingMessages.forEach { m ->
+                if (m.isFtsMessage()) {
+                    MessageFts4Helper.insertOrReplaceMessageFts4(m)
```

Comment:
Maybe it should be handled async

Reply:
I would prefer to put `insert fts` and `insert message` in the same transaction(place).